src/NServiceBus.Core/Recoverability/FirstLevelRetries/FailureInfoStorage.cs

```diff
@@ -0,0 +1,134 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Implements an in memory storage for message processing failures. Each failure contains a number of failed
+    /// processing attempts and the last captured exception.
+    /// </summary>
+    /// <remarks>
+    /// The data structure has fixed size configurable by a constructor parameter. When the data structure reaches
+    /// its maximum size, the least recently used (LRU) message processing failure is removed from the storage.
+    /// </remarks>
+    class FailureInfoStorage
+    {
+        /// <summary>
+        /// Helper storage structure that binds message processing failure information and LRU entry.
+        /// </summary>
+        class FailureInfoNode
+        {
+            public FailureInfoNode(string messageId, ProcessingFailureInfo failureInfo)
+            {
+                FailureInfo = failureInfo;
+                LeastRecentlyUsedEntry = new LinkedListNode<string>(messageId);
+            }
+
+            public ProcessingFailureInfo FailureInfo { get; set; }
+            public LinkedListNode<string> LeastRecentlyUsedEntry { get; }
+        }
+
+        Dictionary<string, FailureInfoNode> failureInfoPerMessage = new Dictionary<string, FailureInfoNode>();
+        LinkedList<string> leastRecentlyUsedMessages = new LinkedList<string>(); 
+        object lockObject = new object();
+
+        int maxElements;
+        
+        /// <summary>
+        /// Initializes a new instance of <see cref="FailureInfoStorage"/>.
+        /// </summary>
+        /// <param name="maxElements">Maximum number of message processing failures stored. Default: 1000.</param>
+        public FailureInfoStorage(int maxElements = 1000)
+        {
+            this.maxElements = maxElements;
+        }
+
+        /// <summary>
+        /// Records a message processing failure. If any failures were previously recorded for a given <paramref name="messageId"/>,
+        /// then the number of failed attempts is incremented and the <paramref name="exception"/> is stored.
+        /// </summary>
+        /// <remarks>
+        /// If the storage already contains maximum allowed number of failure messages, then the least recently used (LRU) item
+        /// is removed from the storage. 
+        /// </remarks>
+        /// <param name="messageId">Unique message identifier.</param>
+        /// <param name="exception">Exception thrown during message processing.</param>
+        public void RecordFailureInfoForMessage(string messageId, Exception exception)
```

Comment:
In theory we can decide to only store the exception when we know that we're going to give up the next time we pull the message up. This would save on memory during retries. Is this something we want to do or YAGNI?